consensus/src/liveness/leader_reputation_test.rs

```diff
@@ -449,51 +451,68 @@ fn test_api() {
-        block_builder.create_block(proposers[0], vec![false, false, false, true, false], vec![]),
-    ];
-    let leader_reputation = LeaderReputation::new(
-        0,
-        proposers.clone(),
-        Box::new(MockHistory::new(1, history)),
-        Box::new(ActiveInactiveHeuristic::new(
-            proposers[0],
-            active_weight,
-            inactive_weight,
-            proposers.len(),
-        )),
-        4,
-    );
-    let round = 42u64;
     // first metadata is ignored because of window size 1
     let expected_weights = vec![
         active_weight,
-        inactive_weight,
-        inactive_weight,
-        active_weight,
-        inactive_weight,
+        inactive_weight * 2,
+        inactive_weight * 3,
+        active_weight * 4,
+        inactive_weight * 5,
     ];
-    let sum = expected_weights.iter().fold(0, |mut s, w| {
-        s += *w;
-        s
-    });
-    let mut state = round.to_le_bytes().to_vec();
-    let chosen_weight = next(&mut state) % sum;
-    let mut expected_index = 0usize;
-    let mut accu = 0u64;
-    for (i, w) in expected_weights.iter().enumerate() {
-        accu += *w;
-        if accu >= chosen_weight {
-            expected_index = i;
-        }
+    let total_weights: u64 = expected_weights.iter().sum();
+
+    let mut selected = [0; 5].to_vec();
+    for epoch in 1..1000 {
+        block_builder.new_epoch();
+        let history = vec![
+            block_builder.create_block(proposers[0], vec![false, true, true, false, false], vec![]),
+            block_builder.create_block(
+                proposers[0],
+                vec![false, false, false, true, false],
+                vec![],
+            ),
+        ];
+        let leader_reputation = LeaderReputation::new(
+            epoch,
+            proposers.clone(),
+            voting_powers.clone(),
+            Box::new(MockHistory::new(1, history)),
+            Box::new(ActiveInactiveHeuristic::new(
+                proposers[0],
+                active_weight,
+                inactive_weight,
+                proposers.len(),
+            )),
+            4,
+        );
+        let round = 42u64;
+        let state = [epoch.to_le_bytes(), round.to_le_bytes()].concat().to_vec();
+
+        let expected_index = choose_index(expected_weights.clone(), state);
+        selected[expected_index] += 1;
+        let unexpected_index = (expected_index + 1) % proposers.len();
+        let output = leader_reputation.get_valid_proposer(round);
+        assert_eq!(output, proposers[expected_index]);
+        assert!(leader_reputation.is_valid_proposer(proposers[expected_index], round));
+        assert!(!leader_reputation.is_valid_proposer(proposers[unexpected_index], round));
+    }
+
+    for i in 0..5 {
+        let p = expected_weights[i] as f32 / total_weights as f32;
+        let expected = (1000.0 * p) as i32;
+        let std_dev = (1000.0 * p * (1.0 - p)).pow(0.5);
```

Comment:
Nice test. Can you add a comment clarifying the intention and the math behind it?